torchgen/api/cpp.py

```diff
@@ -161,7 +172,10 @@ def argumenttype_type(
             else:
                 return NamedCType(binds, BaseCType(symIntArrayRefT))
         elif str(t.elem) == "Tensor":
-            return NamedCType(binds, BaseCType(tensorListT))
+            if structured_type_override:
```

Comment:
Tell me if I'm misunderstanding, but it seems like the main goal of plumbing `structured_type_override` into here, all the way through the various objects that we have (`DispatcherSignature`, `NativeSignature`, etc), is because we now need to know more than just the JIT type in order to decide what C++ code to emit - we also need non-local info about the NativeFunction, specifically whether or not its structured.

I think this is actually exactly the use case for the dynamic scoping that torchgen has here: https://github.com/pytorch/pytorch/blob/684a404defaaa9e55ac905a03d9f6a368f3c45f7/torchgen/local.py#L25. Basically, you can set some thread-local state in the codegen that gets reset per NativeFunction, and query that state here. You could add "is the current NativeFunction structured" here. Then you can avoid having to do a bunch of plumbing / adding extra arguments to signatures around the codegen.

Let me know what you think!

Reply:
Hmm... This sounds interesting. So, for example, inject it inside `native_function_manager` context like:

https://github.com/pytorch/pytorch/blob/684a404defaaa9e55ac905a03d9f6a368f3c45f7/torchgen/context.py#L52-L55

Reply:
yep, exactly. You can add a new local piece of context in addition to `use_const_ref_for_mutable_tensors`, like `is_structured`